### ISSUE TITLE
No alphabetical sort for triggers of same type and number of words

### ISSUE THREAD

**Issue by Dinh-Hung-Tu** [aichaos/rivescript-python#66]
According to the [sorting of triggers](https://www.rivescript.com/wd/RiveScript.html#Sorting-Triggers), 

> For multiple triggers with the same amount of words, sort them by length, and then alphabetically if there are still matches in length. 

However, it seems that the sort by character-length and alphabetical order are not yet implemented. 

An example set of triggers, this can be evaluated from [Try RiveScript Online](https://www.rivescript.com/try):
```
+ * you *
- you_reply
+ * are *
- are_reply
```
Both of them would match the user input `"how are you doing"` but the RiveScript answer will be `you_reply`, the first occurrence in the `.rive` file (because the triggers are stored as list). Change the order of triggers and you will find it matches the first occurrence. The dependence on ordering of templates in such cases violates the sorting rule stated earlier (in alphabetical order).  

Another example of atomic triggers which requires activation of the RS debug mode by setting `rs = RiveScript(debug = True)` to see the sort order of triggers:
```
+ hi
- hi there
+ hello
- hello there
```
The sort order should be: `hello` then `hi` because:

1. Length in terms of characters (`len("hello")=5`, `len("hi")=2`)

2. Alphabetical order. 

However, the order search information shows that sorted list of triggers places `hi` before `hello`
```
You> hello
[RS] Get reply to [localuser] hello
[RS] Checking topic random for any %Previous's.
[RS] There is a %Previous in this topic!
[RS] lastReply: undefined
[RS] Try to match u'hello' against u'hi' (u'^hi$')
[RS] Try to match u'hello' against u'hello' (u'^hello$')
[RS] Found a match!
[RS] Reply: hello there
Bot> hello there
```

**Comment by Dinh-Hung-Tu** [comment]
@kirsle Is this issue a missing feature? I have made an on going fix for this.

**Comment by kirsle** [comment]
Yeah it is: the order of triggers in the input files *should not* affect the order they get sorted at the end.

In your example with `* you *` and `* are *`, they should be sorted with `are > you` because they both have wildcards, the same number of words, and are the same length, so alphabetic sorting should be used, and their order should be deterministic regardless of the order they appeared in the original source file.

**Comment by Dinh-Hung-Tu** [comment]
Noted about this requirement of alphabetical sorting. 

How about the `hi` and `hello` example above? Should the order also be `hello > hi` on the basis of length by character count?

**Comment by Dinh-Hung-Tu** [comment]
One more feature, should we consider the number of special triggers such as trigger with 1 wildcard should be before trigger with 2 wildcards?

**Comment by kirsle** [comment]
> How about the `hi` and `hello` example above? Should the order also be `hello > hi` on the basis of length by character count?

Yeah probably.

> One more feature, should we consider the number of special triggers such as trigger with 1 wildcard should be before trigger with 2 wildcards?

If you find an efficient way to implement this it could be fine. The sorting code is pretty gnarly already. :wink:

The overall idea of the sorting algorithm is to order the most specific triggers on top, so that the wildcards in a short trigger don't end up shadowing a better, longer trigger from matching the message instead. Things like `who is your *` would be preferable to match before a more generic `who is *` would, etc.; the current algorithm mostly works okay except in some edge cases as this ticket mentioned.

(Also, the `{weight}` tag was added as a way to manually hand-tune the priority of triggers for certain edge cases, but should be used sparingly as it actually *drastically* rearranges the sorted list, so when used on short triggers with wildcards it can cause more harm than good).

**Comment by kirsle** [comment]
Fixed in PR #67
